plugins/GSdx/GSFastList.h

```diff
@@ -0,0 +1,162 @@
+/*
+*	Copyright (C) 2017-2017 Alessandro Vetere
+*	Copyright (C) 2007-2009 Gabest
+*	http://www.gabest.org
+*
+*  This Program is free software; you can redistribute it and/or modify
+*  it under the terms of the GNU General Public License as published by
+*  the Free Software Foundation; either version 2, or (at your option)
+*  any later version.
+*
+*  This Program is distributed in the hope that it will be useful,
+*  but WITHOUT ANY WARRANTY; without even the implied warranty of
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+*  GNU General Public License for more details.
+*
+*  You should have received a copy of the GNU General Public License
+*  along with GNU Make; see the file COPYING.  If not, write to
+*  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA USA.
+*  http://www.gnu.org/copyleft/gpl.html
+*
+*/
+
+#pragma once
+
+template <class T>
+struct Element {
+	T  data;
+	uint16 next_index;
+	uint16 prev_index;
+};
+
+template <class T, uint16 INITIAL_SIZE = 4>
+class FastList {
+private:
+	// The last Element<T> of the list has next_index equal to 0
+	// m_buffer[0] is the head of the list, and m_buffer has m_size elements
+	Element<T>* m_buffer;
+	uint16 m_size;
+	uint16 m_free_indexes_stack_top;
+	// m_buffer indexes that are free to be used are stacked here [END - 1]
+	uint16* m_free_indexes_stack;
+
+public:
+	FastList() {
+		m_buffer = nullptr;
+		Clear();
+	}
+
+	~FastList() {
+		_aligned_free(m_buffer);
+	}
+
+	void Clear() {
+		// Initialize END
+		m_size = INITIAL_SIZE;
+		// Initialize arrays
+		_aligned_free(m_buffer);
+		m_buffer = (Element<T>*)_aligned_malloc(m_size * sizeof(Element<T>) + (m_size - 1) * sizeof(uint16), 64);
+		m_free_indexes_stack = (uint16*)&m_buffer[m_size];
+		// Initialize head
+		m_buffer[0].data = nullptr;
+		m_buffer[0].next_index = End();
+		m_buffer[0].prev_index = 0;
+		// Stack top is 0, bottom is END - 2
+		m_free_indexes_stack_top = 0;
+		// Index 0 reserved for head
+		for (uint16 i = 0; i < m_size - 1; i++) {
+			m_free_indexes_stack[i] = i + 1;
+		}
+	}
+
+	__forceinline uint16 First() const {
```

Comment:
The fact is that the C++11 range loops would be only used in the part of the code where they are already used now (which is only when looping on the lists to destroy Source objects if I remember well), in all the other parts of the code looping using iterators is needed because you need to have the iterator to the element to perform operations (like MoveFront or Erase).
So introducing iterators would only shorten one usage of the list, while leaving unchanged (ok, instead of using the helper functions you can do ++ operator) most of the code.